workflow.go

```diff
@@ -625,3 +626,43 @@ const DefaultVersion Version = -1
 func GetVersion(ctx Context, changeID string, minSupported, maxSupported Version) Version {
 	return getWorkflowEnvironment(ctx).GetVersion(changeID, minSupported, maxSupported)
 }
+
+// SetQueryHandler sets the query handler to handle workflow query. The queryType specify which query type this handler
+// should handle. The handler must be a function that returns 2 values. The first return value must be a serializable
+// result. The second return value must be an error. The handler function could receive any number of input parameters.
+// All the input parameter must be serializable. You should call cadence.SetQueryHandler() at the beginning of the workflow
+// code.
+// Example of workflow code that support query type "current_state":
+//  func MyWorkflow(ctx cadence.Context, input string) error {
+//    currentState := "started" // this could be any serializable struct
+//    err := cadence.SetQueryHandler(ctx, "current_state", func() (string, error) {
+//      return currentState, nil
+//    })
+//    if err != nil {
+//      currentState = "failed to register query handler"
```

Comment:
Should we support querying closed and failed workflows? And what is the semantic of it?
If not this example doesn't make sense as it updates currentState before closing the workflow.

Reply:
Technically, we could support. The current implementation does support it. The use case is that user don't need to check the open/closed state of a workflow before sending out query. User could also use query to return workflow result or activity result if they wish to. (However, extending visibility API to return workflow result would be much cheaper.) I don't know, I don't feel strong about this. We can discuss again on this.